### ISSUE TITLE
[Bookings] Booking form runs one unindexed date-range query per candidate time slot — 17s+ form renders on stores with booking history

### ISSUE THREAD

**Issue by reddjett** [woocommerce/woocommerce#66523]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

On stores with a few years of booking history, the front-end booking form for customer-defined duration products becomes unusably slow: ~17 seconds to render the start-time dropdown and ~40 seconds for the end-time dropdown (max_duration 100). This worsens gradually over a store's life, so it presents as "Bookings gets slower every year."

Profiling the full call chain shows the root cause is an N+1 query pattern:

WC_Bookings_Ajax::get_time_blocks_for_date → WC_Booking_Form::get_time_slots_html
→ get_start_time_html() calls get_end_times( ..., $check = true ) once per available start block (~48/day for 30-minute blocks)
→ each check calls wc_bookings_get_total_available_bookings_for_range() → get_blocks_availability() → WC_Booking_Data_Store::get_bookings_in_date_range()
get_end_time_html() repeats the same per-candidate check for every duration step up to max_duration.

get_bookings_in_date_range() costs ~340ms per call on our dataset: it self-joins wp_postmeta on _booking_start / _booking_end and compares meta_value (longtext) with no usable index, so every call scans the booking meta subset. ~48 candidate slots × ~340ms ≈ 17s per click.

The book_ts_ / book_st_ transient caches mask this only for repeat views of an identical date/range; every fresh date, product, or cache invalidation pays full price. (Caution for anyone reproducing: WC_Bookings_Cache persists per-range results, so benchmarks on a previously-touched range appear ~100x faster than a cold range.)

Measured impact

OperationStockWith postmeta indexget_bookings_in_date_range() — single call, cold range~340ms~4msget_time_slots_html() — 48 blocks, cold17.1s0.26s

Workaround we applied in production:

sqlALTER TABLE wp_postmeta ADD INDEX bookings_meta_key_value (meta_key(64), meta_value(20));

### Expected behavior

Expected behavior
Calendar and both time dropdowns should resolve in well under a second regardless of how many past bookings the store has accumulated. Booking history that can never conflict with a future date should not affect the time to render availability.

### Actual behavior

On a store with ~2,800 booking records (accumulated since 2022), opening a bookable product and picking a date:

The availability calendar takes 3–8 seconds to resolve which days are available.
After choosing a date, the start-time dropdown takes 15–20 seconds to populate.
After choosing a start time, the end-time dropdown takes ~40 seconds to populate (max_duration 100).

Every step gets slower as booking history grows, so the form degrades a little more each year. Repeat views of the exact same date are fast (transient cache), but any fresh date/product pays the full cost — which is most real customer traffic.

### Steps to reproduce

Store with ~2,500+ wc_booking posts (~250k total postmeta rows) — the bookings-generator repo can produce this.
A bookable product: customer-defined duration, 30-minute blocks, min 2 / max 100 duration, 12-month max bookable window, no resources/persons.
Front end, logged out: open the product, pick any date that hasn't been picked recently (cold cache).
Time the admin-ajax.php?action=wc_bookings_get_blocks request and the subsequent get_end_time_html request.

### WordPress Environment

<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 10.9.4
Action Scheduler Version: ✔ 3.9.3
Log Directory Writable: ✔
WP Version: 7.0.1
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: ✔
WP Cron: –
WP Environment Type: production
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache
Server Architecture: Linux 3.10.0-1160.119.1.el7.tuxcare.els19.x86_64 x86_64
PHP Version: 8.3.31
PHP Post Max Size: 128 MB
PHP Time Limit: 600
PHP Max Input Vars: 1000
cURL Version: 8.17.0
OpenSSL/1.1.1w

SUHOSIN Installed: –
MySQL Version: 10.3.39-MariaDB
Max Upload Size: 10 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

aiosrs-schema: 1
amp_cta_bar: 1
attachment: 272
bookable_person: 18
custom_css: 1
divi_bars: 3
et_pb_layout: 18
follow_up_email: 7
ig_campaign: 4
ig_message: 6
jetpack_migration: 2
kwlogos: 4
nav_menu_item: 12
oembed_cache: 10
page: 27
post: 7
product: 64
revision: 19
shop_coupon: 10
shop_order: 3402
shop_order_refund: 148
vfb_entry: 1
wc_booking: 2818
wc_zapier_feed: 2
wp_global_styles: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (48) ###

Advanced Cron Manager: by BracketSpace – 2.7.0
Advanced Reporting For Woocommerces: by phoeniixx – 3.0
Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.7
Bulk Delete: by WebFactory Ltd – 6.11
Burst Statistics - Privacy-Friendly Analytics for WordPress: by Burst Statistics - Stats & Analytics for WordPress – 3.6.0.1
CallRail Phone Call Tracking: by CallRail
Inc. – 0.5.3

Code Snippets: by Code Snippets Pro – 3.9.6
Coinbase Commerce: by Coinbase Commerce – 1.4.1
Divi Bars: by Divi Life — Tim Strifler – 1.9.6
Divi Font Awesome: by Alex Brinkman – 1.2.2
Download Plugins and Themes in ZIP from Dashboard: by WPFactory – 1.9.9
Duplicate Page: by mndpsingh287 – 4.5.9
GTM4WP - A Google Tag Manager (GTM) plugin for WordPress: by Thomas Geiger – 1.22.3
Email Log: by WebFactory Ltd – 2.62
Enable Media Replace: by ShortPixel – 4.2.2
EWWW Image Optimizer: by Exactly WWW – 8.7.3
Icegram Engage – Popups, Optins, CTAs & Lead Generation: by icegram – 3.1.42
Mailchimp for WooCommerce: by Mailchimp – 6.1.1
Media Library Assistant: by David Lingren – 3.38
SOGO Add Script Header Footer: by orenhav (SOGO) – 3.9
Optimus Divi Logo Slider Module: by Optimus Divi – 1.2.8
Pixel Manager for WooCommerce (Premium): by SweetCode – 1.58.5
Really Simple Security: by Really Simple Security – 9.6.1
Send E-mails with Mandrill: by Miller Media ( Matt Miller ) – 1.6.2
Easy Updates Manager: by Easy Updates Manager Team – 9.0.21
UpdraftPlus - Backup/Restore: by TeamUpdraft
DavidAnderson – 1.26.5

Advanced Order Export For WooCommerce: by AlgolPlus – 4.1.0
Preview E-mails for WooCommerce: by Digamber Pradhan – 2.2.14
WooCommerce Advanced Bulk Edit: by George Iron & Yas G. – 5.1
WooCommerce Advanced Notifications: by WooCommerce – 1.6.9
WooCommerce Bookings: by WooCommerce – 3.7.0
WooCommerce Checkout Field Editor: by WooCommerce – 1.7.25
WooCommerce Colors: by WooThemes – 1.0.7
Follow-Up Emails: by WooCommerce – 4.9.51
WooCommerce Stripe Gateway: by Stripe – 10.8.3
WooCommerce Google Analytics Pro: by SkyVerge – 2.2.3
WooCommerce PayPal Payments: by PayPal – 4.1.1
WooCommerce Product Add-Ons: by WooCommerce – 8.3.0
WooCommerce Tax: by WooCommerce – 3.6.7
WooCommerce Smart Coupons: by StoreApps – 9.63.0
WooCommerce Zapier: by Tectalic (formerly OM4 Software) – 2.11.1
WooCommerce: by Automattic – 10.9.4
Yoast SEO Premium: by Team Yoast – 26.4
Yoast SEO: by Team Yoast – 28.0
ManageWP - Worker: by GoDaddy – 4.9.34
WP-Optimize - Clean, Compress, Cache: by TeamUpdraft
DavidAnderson – 4.6.0

WP STAGING WordPress Backup Plugin - Backup Duplicator & Migration: by WP-STAGING
WPStagingBackup – 4.9.2

Yoast SEO: WooCommerce: by Team Yoast – 16.7

### Inactive Plugins (3) ###

All-in-One WP Migration and Backup: by ServMask – 7.106
VFB Pro: by Matthew Muro – 3.4.6
VFB Pro - Form Designer: by Matthew Muro – 2.0.5

### Dropin Plugins () ###

advanced-cache.php: advanced-cache.php

### Must Use Plugins (6) ###

Burst REST API Optimizer: by Burst Statistics – 1.0.0
effectio.php: by Effectio Team –
ET Support Center :: Must-Use Plugins Autoloader: by Elegant Themes –
VPP Bookings Availability Cache (v2): by Claude (for Benjamin)
2026-07-12 –

VPP Bookings Fast Time Slots: by Claude (for Benjamin)
2026-07-12 –

WP STAGING Optimizer: by WP STAGING – 1.6.0

### Settings ###

Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: booking (booking)
external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔
Enforce Approved Product Download Directories: –
HPOS feature enabled: –
Order datastore: WC_Order_Data_Store_CPT
HPOS data sync enabled: –
Enabled Features: analytics
rate_limit_checkout
marketplace
order_attribution
site_visibility_badge
blueprint
point_of_sale
push_notifications


### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 14 MB

### WC Pages ###

Shop base: #15 - /shop/
Cart: #16 - /cart/ -  Contains the [woocommerce_cart] shortcode
Checkout: #17 - /checkout/ -  Contains the [woocommerce_checkout] shortcode
My account: #18 - /my-account/ -  Contains the [woocommerce_my_account] shortcode
Terms and conditions: ❌ Page not set

### Theme ###

Name: Divi
Version: 4.27.7
Author URL: http://www.elegantthemes.com
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
Theme type: Classic theme
WooCommerce Support: ✔

### Templates ###

Override: wp-content/plugins/woocommerce-checkout-field-editor/templates/emails/email-addresses.php ❌ Version 3.5.4 is out of date. The core version is 10.6.0
Override: wp-content/plugins/woocommerce-checkout-field-editor/templates/emails/plain/email-addresses.php ❌ Version 3.4.0 is out of date. The core version is 8.6.0
Override: wp-content/plugins/woocommerce-checkout-field-editor/templates/order/order-details-customer.php ❌ Version 3.4.4 is out of date. The core version is 8.7.0

### WooCommerce Stripe Payment Gateway ###

Version: 10.8.3
Account ID: acct_1Cqt9GG8vn4hJJqk
Test Mode Enabled: ✔
					No
OAuth Connected: ✔
					Yes
Sync Enabled: ✔
					Yes
Optimized Checkout Enabled: ✔
						Yes
Enabled Payment Methods: card,link
Express Checkout: ✔
					Enabled (product,cart)
Auth and Capture: ✔
					Yes
Logging: No

### WooCommerce PayPal Payments ###

Onboarded: ✔
Branded only: –
Shop country code: US
WooCommerce currency supported: ✔
Advanced Card Processing available in country: ✔
Pay Later messaging available in country: ✔
Webhook status: ✔
PayPal Vault enabled: –
ACDC Vault enabled: –
Logging enabled: –
Reference Transactions: ✔
Used PayPal Checkout plugin: ✔
Subscriptions Mode: Disabled
PayPal Shipping Callback: –
Apple Pay: –
Google Pay: ✔
Fastlane: ✔

### Admin ###

Enabled Features: activity-panels
analytics
analytics-scheduled-import
product-block-editor
experimental-iapi-mini-cart
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
onboarding
onboarding-tasks
pattern-toolkit-full-composability
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
printful
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store

Disabled Features: product-data-views
experimental-blocks
coming-soon-newsletter-template
minified-js
products-catalog-api
settings-ui
async-product-editor-category-field
use-wp-horizon
rest-api-v4
order-detail-redesign
product-variations-classic-redesign

Daily Cron: ✔ Next scheduled: 2026-07-12 19:49:04 -07:00
Options: ✔
Notes: 298
Onboarding: completed

### Action Scheduler ###

Complete: 3,573
Oldest: 2026-06-26 18:11:04 -0700
Newest: 2026-07-12 11:02:02 -0700

Failed: 182
Oldest: 2021-03-14 04:02:41 -0700
Newest: 2026-05-07 12:20:11 -0700

Pending: 18
Oldest: 2026-07-12 12:56:03 -0700
Newest: 2026-07-22 18:22:08 -0700


### Smart Coupons related settings ###

⠀* Number of coupons to show: 5
⠀* Number of characters in auto-generated coupon code: 
⠀* Valid order status for auto-generating coupon: Processing | Completed
⠀* Automatic deletion: ❌ No
⠀* Coupon emails: ✔ Yes
⠀* Printing coupons: ✔ Yes
⠀* Sell gift cards at less price?: ❌ No
⠀* Use gift card applied in first subscription order for subsequent renewals until credit reaches zero: ✔ Yes
⠀* Allow sending of coupons to others: ✔ Yes
⠀* Allow schedule sending of coupons?: 
⠀* Combine emails: 
⠀* Auto generated coupon email: ✔ Yes
⠀* Combined auto generated coupons email: ❌ No
⠀* Acknowledgement email: ✔ Yes
⠀* Include tax in the amount of the generated gift card: 
⠀* Deduct credit/gift before doing tax calculations: ❌ No
⠀* Gift Card discount is inclusive of tax: ❌ No
⠀* Enable taxes: ❌ No
⠀* Prices entered with tax: ❌ No
⠀* Display prices in the shop: ❌ Excl
⠀* Display prices during cart and checkout: ❌ Excl
⠀* Rounding: ❌ No
⠀* Display tax totals: Single
⠀* Enable the use of coupon codes: ✔ Yes
⠀* Calculate coupon discounts sequentially: ❌ No
⠀* Account endpoints > Coupons: wc-smart-coupons
⠀* WooCommerce Account Connected: ✔ Yes

### WooCommerce Zapier ###

Database Version: ✔ 20
Detailed Logging Enabled: No
Legacy Mode Disabled: ❌ No
SSL: ✔ Yes
REST API Authentication Key(s): ❌ 0
Task History Record Count: ✔ 1681
Webhook #1: WooCommerce Zapier
- Status: active
- Trigger: order.created
- Delivery Count: 1681
- User: VPPadmin
Legacy Cron Tasks: ✔ 0
Active Legacy Feeds: ❌ 2 active Legacy Feed(s) need to be migrated before October 14
2024. Migrate Your Zaps Now

Legacy Feed #1: VPP Woo to Sheets
- Trigger: New Order
- Webhook URL: https://hooks.zapier.com/hooks/catch/2418849/vasopr/
Legacy Feed #2: New Order Party Bus
- Trigger: New Order
- Webhook URL: https://hooks.zapier.com/hooks/catch/2418849/vvgouj/

### Status report information ###

Generated at: 2026-07-12 11:06:55 -07:00
```
</details>

### Isolating the problem

- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by xristos3490** [comment]
Hey @reddjett! Thanks for the detailed report; much appreciated!

I can confirm this is an issue with the WooCommerce Bookings extension, and we have opened a separate issue to track this work. I'll close this in the WooCommerce repo, but we'll provide resolution details once this is fixed. 

Thanks again!
